src/certificates_observer.py

```diff
@@ -27,6 +27,9 @@ def __init__(self, charm: CharmBaseWithState):
         super().__init__(charm, RELATION_NAME)
         self._charm = charm
         self.private_key = certificates.generate_private_key().decode()
+        self.csr = certificates.generate_csr(
+            private_key=self.private_key.encode(), subject=self._charm.unit.name
+        )
```

Comment:
You can't generate a new CSR for every event, you need to generate it for the first time, or when the certificate needs to be renewed and store it in a peer relation, or retrieve it (the last one ?) from the certificate relation

Reply:
That means tracking state in a peer relation of something similar. I can do that in a follow up PR, that this will fix the functional issue regardless

Reply:
Fixed with a secret. CSRs should be idempotent now